File: client/src/dto/PartialMatchActionDto.ts
import { ActionType } from "../enums/actionType";
import { CoordinatesDto } from "./CoordinatesDto";

export interface PartialMatchActionDto {
    player1: boolean;
    type: ActionType;
    originatingCellCoords?: CoordinatesDto;
    impactedCoords: CoordinatesDto[];
}

File: server/dto/partial_match_action_dto.py
from dataclasses import dataclass
from typing import TYPE_CHECKING

from dto.base_dto import BaseDto
from dto.coordinates_dto import CoordinatesDto

if TYPE_CHECKING:
    from dto.server_only.match_action_dto import ActionType, MatchActionDto


@dataclass
class PartialMatchActionDto(BaseDto):
    player1: bool
    type: "ActionType"
    originatingCellCoords: CoordinatesDto
    impactedCoords: tuple[CoordinatesDto]

    @staticmethod
    def from_match_action_dto(match_action_dto: "MatchActionDto"):
        if not match_action_dto:
            return None

        return PartialMatchActionDto(
            match_action_dto.player1,
            match_action_dto.type,
            match_action_dto.originatingCellCoords,
            match_action_dto.impactedCoords,
        )
